tiledb/sm/c_api/tiledb.cc

```diff
@@ -4116,17 +4114,15 @@ int32_t tiledb_array_alloc(
   }
 
   // Allocate an array object
-  (*array)->array_ =
-      new (std::nothrow) tiledb::sm::Array(uri, ctx->ctx_->storage_manager());
-  if ((*array)->array_ == nullptr) {
-    delete *array;
-    *array = nullptr;
+  try {
+    (*array)->array_ = make_shared<tiledb::sm::Array>(
+        HERE(), uri, ctx->ctx_->storage_manager());
+  } catch (...) {
```

Comment:
Catch `std::bad_alloc` here so that it matches the message below rather than catching all exceptions.